src/marlin/marlin_pst13_pc/data_structures.rs

```diff
@@ -0,0 +1,572 @@
+use crate::{BTreeMap, Vec};
+use crate::{
+    PCCommitterKey, PCPreparedVerifierKey, PCProof, PCRandomness, PCUniversalParams, PCVerifierKey,
+};
+use ark_ec::PairingEngine;
+use ark_ff::{ToBytes, Zero};
+use ark_poly::MVPolynomial;
+use ark_std::{
+    io::{Read, Write},
+    marker::PhantomData,
+    ops::{Add, AddAssign, Index},
+};
+
+use ark_serialize::{CanonicalDeserialize, CanonicalSerialize, SerializationError};
+use rand_core::RngCore;
+
+/// `UniversalParams` are the universal parameters for the MarlinPST13 scheme.
+#[derive(Derivative)]
+#[derivative(Default(bound = ""), Clone(bound = ""), Debug(bound = ""))]
+pub struct UniversalParams<E, P>
+where
+    E: PairingEngine,
+    P: MVPolynomial<E::Fr>,
+    P::Point: Index<usize, Output = E::Fr>,
+{
+    /// Contains group elements corresponding to all possible monomials with
+    /// `num_vars` and maximum degree `max_degree` evaluated at `\beta`
+    pub powers_of_g: BTreeMap<P::Term, E::G1Affine>,
+    /// `\gamma` times the generater of G1
+    pub gamma_g: E::G1Affine,
+    /// Group elements of the form `{ \beta_i^j \gamma G }`, where `i` ranges
+    /// from 0 to `num_vars-1` and `j` ranges from `1` to `max_degree+1`.
+    pub powers_of_gamma_g: Vec<Vec<E::G1Affine>>,
+    /// The generator of G2.
+    pub h: E::G2Affine,
+    /// Group elements of the form `{ \beta_i H }`, where `i` ranges from 0 to `num_vars-1`
+    pub beta_h: Vec<E::G2Affine>,
+    /// The generator of G2, prepared for use in pairings.
+    #[derivative(Debug = "ignore")]
+    pub prepared_h: E::G2Prepared,
+    /// Group elements of the form `{ \beta_i H }`, where `i` ranges from 0 to `num_vars-1`,
+    /// prepared for use in pairings
+    #[derivative(Debug = "ignore")]
+    pub prepared_beta_h: Vec<E::G2Prepared>,
+    /// The number of variables `self` is initialized for
+    pub num_vars: usize,
+    /// The maximum degree supported by `self`
+    pub max_degree: usize,
+}
+
+impl<E, P> CanonicalSerialize for UniversalParams<E, P>
```

Comment:
Is there a reason the `derive` doesn't work here?

Reply:
Due to `G2Prepared`. Ideally, we need `#[derivative(CanonicalSerialize = "ignore")]` so some of the entries are ignored during serialization. Nevertheless, for such, we still need to implement CanonicalDeserialize.